libr/include/r_arch.h

```diff
@@ -0,0 +1,100 @@
+/* radare2 - LGPL - Copyright 2020 - pancake */
+
+#ifndef R2_ARCH_H
+#define R2_ARCH_H
+
+#include <r_util.h>
+#include <r_bind.h>
+#include <r_io.h> // Just for RIOBind which is needed for some plugins
+
+typedef int RArchBits;
+typedef int RArchEndian;
+
+typedef enum r_arch_decode_options_t {
+	R_ARCH_OPTION_CODE = 1 << 0, // 1,
+	R_ARCH_OPTION_SIZE = 1 << 1, // 2,
+	R_ARCH_OPTION_ANAL = 1 << 2, //4,
+	R_ARCH_OPTION_ESIL = 1 << 3, //8,
+	R_ARCH_OPTION_OPEX = 1 << 4, //16,
+	R_ARCH_OPTION_DESC = 1 << 5, // 32, // instruction description
+} RArchOptions;
+
+typedef struct r_arch_t {
+	struct r_arch_plugin_t *cur;
+	RArchEndian endian;
+	RArchBits bits;
+	char *cpu;
+	RList *plugins;
+	RIOBind iob;
+	size_t syntax;
+} RArch;
+
+typedef struct r_arch_info_t {
+	size_t align;
+} RArchInfo;
+
+typedef struct r_arch_instruction_t {
+	ut64 addr;
+	size_t size; // inherit from thisi.data.len?
+	ut32 opid; // opcode identifier
+	ut64 type; // opcode type
+	RStrBuf code; // disasm
+	RStrBuf data; // bytes
+	RStrBuf esil; // emulation
+	RStrBuf opex; // analysis
+	size_t syntax; // used for disasm
+	RVector dest; // array of destinations
+} RArchInstruction;
+
+
+typedef bool (*RArchEncodeCallback)(RArch *a, RArchInstruction *ins, RArchOptions options);
+typedef bool (*RArchDecodeCallback)(RArch *a, RArchInstruction *ins, RArchOptions options);
+typedef char * (*RArchRegistersCallback)(RArch *a);
+typedef void (*RArchInfoCallback)(RArch *a);
+
+typedef struct r_arch_plugin_t {
+	const char *name;
+	const char *arch;
+	const char *cpus;
+	const char *features;
+	RArchBits bits;
+	RArchEndian endian;
+	// copyright
+	const char *author;
+	const char *desc;
+	const char *license;
+	const char *version;
+	// callbacks
+	RArchEncodeCallback encode;
+	RArchDecodeCallback decode;
+	RArchRegistersCallback registers;
+	RArchInfoCallback info; // dynamic info
+	bool (*init)(void *user);
+	bool (*fini)(void *user);
+} RArchPlugin;
+
+R_API RArchInstruction *r_arch_instruction_new();
+R_API void r_arch_instruction_free(RArchInstruction *ins);
+R_API void r_arch_instruction_init(RArchInstruction *ins);
+R_API void r_arch_instruction_init_data(RArchInstruction *ins, ut64 addr, const ut8 *buf, size_t len);
+R_API void r_arch_instruction_init_code(RArchInstruction *ins, ut64 addr, const char *opstr);
+R_API void r_arch_instruction_fini(RArchInstruction *ins);
+
+R_API RArch *r_arch_new();
+R_API void r_arch_free(RArch *arch);
+
+R_API bool r_arch_encode(RArch *a, RArchInstruction *ins, RArchOptions opt);
+R_API bool r_arch_decode(RArch *a, RArchInstruction *ins, RArchOptions opt);
```

Comment:
same here. `r_arch_decode` always use `ins.data`, why not making the data just a argument of this function. Something like
```C
ins = r_arch_decode (arch, (const ut8 *)"\x90\x90", opt);
```

Reply:
thats wrong. you should pass the length too, and this was my initial design as i said. but i want to encode decoded instructions too, whta i have in mind its not just bytes <-> disasm.

Reply:
you also have to pass the program counter for that call to work, so you  end up with 5 arguments, also:

* using a struct you don't have to copy the bytes into the struct because when you decode you expect those bytes to be available in the struct to show them in hexdump: "90 nop", this 0x90 must be stored in the struct of the instruction too, so you end up filling the struct soon or late. what you are doing is what RAsm.disasemble() does and imho this is a higher level design on top of RArch

Reply:
I don't see `PC` being set in the unit tests.
Also, if you have other usecases in mind please implement them as unit tests to show them off, so we can evaluate those as well (you can mark the tests as broken if those things are not implemented yet).
With regard to copying bytes, you have anyway to do that sooner or later, in both approaches.
With your approach, you copy bytes before calling r_arch_decode, with my approach you copy bytes inside r_arch_decode. I don't see a difference in that regard.

Reply:
It's ok to bundle the input data in one struct. But encoded and decoded instruction should be clearly separated. If you want to have only single parameters for each, you can use two structs `RArchDecodedInstruction` and `RArchEncodedInstruction` but not everything in one.